psdisk/psdisk.c

```diff
@@ -117,199 +116,220 @@ static void psdisk_showPartsTable(void)
 	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
-		printf(PARTITION_DATA_ALIGMENT, node->pHeader.name, node->pHeader.offset, node->pHeader.offset + node->pHeader.size,
-			   node->pHeader.size / psdisk_common.mem.sectorSize, node->pHeader.size, psdisk_getTypeName(node->pHeader.type));
+		printf(PARTITION_DATA_ALIGMENT, node->part.name, node->part.offset, node->part.offset + node->part.size,
+			   node->part.size / psdisk_common.blksz, node->part.size, psdisk_getTypeName(node->part.type));
 	}
 	printf("\n");
 }
 
 
-/* Callbacks to ptable library */
+/* Partition table read/write functions */
 
-static ssize_t psdisk_readImg(unsigned int addr, void *buff, size_t bufflen)
+static ptable_t *psdisk_readImg(void)
 {
-	int offs;
+	ptable_t *ptable;
+	uint32_t count, size;
 
-	if (PTABLE_WITH_OFFSET(psdisk_common.opts))
-		offs = addr;
-	else
-		offs = addr - (psdisk_common.mem.memSize - psdisk_common.mem.sectorSize);
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return NULL;
+	}
 
-	if (fseek(psdisk_common.file, offs, SEEK_SET) != 0) {
-		fprintf(stderr,"Cannot change file position, err: %s.\n", strerror(errno));
-		return -1;
+	if (fread(&count, 1, sizeof(count), psdisk_common.file) != sizeof(count)) {
+		return NULL;
 	}
 
-	return fread(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
-}
+#if __BYTE_ORDER != __LITTLE_ENDIAN
+	count = le32toh(count);
+#endif
 
+	size = ptable_size(count);
+	if (size > psdisk_common.blksz) {
+		return NULL;
+	}
 
-static ssize_t psdisk_writeImg(unsigned int addr, const void *buff, size_t bufflen)
-{
-	ssize_t res;
-	char *wBuff;
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return NULL;
+	}
 
-	if (bufflen > psdisk_common.mem.sectorSize) {
-		fprintf(stderr,"Partition table exceeds sector size. Reduce number of partitions or increase sector size.");
-		return -1;
+	ptable = malloc(size);
+	if (ptable == NULL) {
+		return NULL;
 	}
 
-	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
-		fprintf(stderr,"Cannot change file position, err: %s.\n", strerror(errno));
-		return -1;
+	if (fread(ptable, 1, size, psdisk_common.file) != size) {
+		free(ptable);
+		return NULL;
 	}
 
-	if ((wBuff = (char *)malloc(psdisk_common.mem.sectorSize)) == NULL) {
-		fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
-		return -1;
+	if (ptable_deserialize(ptable, psdisk_common.memsz, psdisk_common.blksz) < 0) {
+		free(ptable);
+		return NULL;
 	}
 
-	memset(wBuff, 0xff, psdisk_common.mem.sectorSize);
+	return ptable;
+}
 
-	if (PTABLE_WITH_OFFSET(psdisk_common.opts)) {
-		while (addr > 0) {
-			fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
-			addr -= psdisk_common.mem.sectorSize;
-		}
+
+static int psdisk_writeImg(ptable_t *ptable)
+{
+	uint32_t size;
+
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return -1;
 	}
+	size = ptable_size(ptable->count);
```

Comment:
Nitpick - `psdisk_readImg()` do check size against `blksz`, so perhaps `psdisk_writeImg()` should do too for symmetry?

Reply:
Ack